### PR TITLE
[Features]Quantize pipeline

### PR DESCRIPTION
**Pull request by HIT-cwh**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Please describe the motivation of this PR and the goal you want to achieve through this PR.

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward compatibility of the downstream repositories?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here and update the documentation.

## Checklist

**Before PR**:

- [ ] Pre-commit or other linting tools are used to fix the potential lint issues.
- [ ] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [ ] The modification is covered by complete unit tests. If not, please add more unit tests to ensure the correctness.
- [ ] The documentation has been modified accordingly, like docstring or example tutorials.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

File: mmrazor/engine/runner/quantization_loops.py [hunk 1]
```diff
@@ -45,62 +48,154 @@ def __init__(
...
             self.run_epoch()
 
             self._decide_current_val_interval()
             if (self.runner.val_loop is not None
                     and self._epoch >= self.val_begin
                     and self._epoch % self.val_interval == 0):
+                # observer disabled during evaluation
+                self.runner.model.state = (False, True)
+                self.runner.model.sync_param()
                 self.runner.val_loop.run()
```
Comment (humu789) [inline mmrazor/engine/runner/quantization_loops.py:78]: 在训练结束后应该再进行 convert + val，convert 成inference模式再测精度更能准确描述量化模型的实际精度

**Review by humu789 — CHANGES_REQUESTED** [review]

File: configs/quantization/qat/lsq_resnet18_8xb32_in1k.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    #     num_classes=1000,
+    #     # RGB format normalization parameters
+    #     mean=[123.675, 116.28, 103.53],
+    #     std=[58.395, 57.12, 57.375],
+    #     # convert image from BGR to RGB
+    #     to_rgb=True,
+    # ),
+    architecture=resnet,
+    quantizer=dict(
+        type='TensorRTQuantizer',
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply configs/quantization/qat/lsq_resnet18_8xb32_in1k.py]: 现在TensorRTQuantizer和CustomQuantizer的差距只差了个qconfig

File: mmrazor/engine/runner/quantization_loops.py [hunk 3]
```diff
@@ -122,6 +217,79 @@ class PTQLoop(TestLoop):
...
+        super().__init__(runner, dataloader, evaluator, fp16)
+
+    def run(self) -> None:
+        """Launch test."""
+        self.runner.call_hook('before_test')
+        self.runner.call_hook('before_test_epoch')
+        self.runner.model.eval()
+        self.runner.model.state = (1, 0)
+
+        for idx, data_batch in enumerate(self.dataloader):
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply mmrazor/engine/runner/quantization_loops.py:234]: 感觉ptq还得再讨论下怎么设计接口，目前主要支持了qat

File: mmrazor/engine/runner/quantization_loops.py [hunk 4]
```diff
@@ -122,6 +217,79 @@ class PTQLoop(TestLoop):
...
+            self.run_iter(idx, data_batch)
+
+        # compute metrics
+        metrics = self.evaluator.evaluate(len(self.dataloader.dataset))
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply mmrazor/engine/runner/quantization_loops.py:238]: 目前convert后的模型inference会报错，还在debug

File: mmrazor/models/algorithms/quantization/base.py [hunk 5]
```diff
@@ -33,74 +38,95 @@ class GeneralQuant(BaseAlgorithm):
...
         super().__init__(architecture, data_preprocessor, init_cfg)
+        if pretrained_ckpt:
+            _ = load_checkpoint(self.architecture, pretrained_ckpt)
+            self.architecture._is_init = True
         self.quantizer = MODELS.build(quantizer)
-        self.observers_enabled = True
-        self.fake_quants_enabled = True
-        self.gen_graphs(self.architecture)
+        self._observers_enabled = True
+        self._fake_quants_enabled = True
```
[the remark this answers is not in the document]
Comment (HIT-cwh) [reply mmrazor/models/algorithms/quantization/base.py:57]: _observers_enabled _fake_quants_enabled是不是应该只能通过GeneralQuant来改呢？通过state.setter

File: configs/quantization/qat/lsq_resnet18_8xb32_in1k.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (humu789) [reply configs/quantization/qat/lsq_resnet18_8xb32_in1k.py]: 是的，我的意思后面TensorRTQuantizer会有较大修改，这里最好先用CustomQuantizer，你这个是用来验证pipeline的，优先保证实验稳定可复现，避免后续改动造成影响

File: mmrazor/engine/runner/quantization_loops.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (humu789) [reply mmrazor/engine/runner/quantization_loops.py:219]: 那这个可以留着后面再改的

File: mmrazor/engine/runner/quantization_loops.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (humu789) [reply mmrazor/engine/runner/quantization_loops.py:223]: 那我先把这个问题弄成待解决项

File: mmrazor/models/algorithms/quantization/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (humu789) [reply mmrazor/models/algorithms/quantization/base.py:61]: 这个得先改了，原来的设计是通过algorithm这个层面修改state的。
现在的主要问题是 _observers_enabled -> observers_enabled；修改state的动作实际是在基类里通过self.observers_enabled属性实现的
_fake_quants_enabled同理

**Review by humu789 — APPROVED** [review]
整体没有什么问题，考虑到整体进度，有些小的comments可以后续再修。

**Pull request merged**